Byte skipping between ID3 and MP3 blocks

localthomas opened the pull request:
As discussed in #28, junk bytes between initial metadata blocks and following MP3 frames are skipped with this implementation change.

The new function `search_for_frame_sync` uses a reader to search for the first appearence of frame sync bits. This function is then used for probing the file type (`Probe::guess_inner`) and reading MP3 data (`mp3::read::read_from`). All three additions are checked by three new tests. The tests for the new function and the `Probe` are in the same module as the implementation, while the `mp3::read_from` test is separted into the `tests` and uses an example `*.mp3` file.

Please comment on the positioning of the tests and the changes on the two existing functions.

---

I am a bit confused by the `Probe::guess_inner` function: 
* Why does it return `Result<Option<FileType>>`?
* What is the logical difference between the return values `Ok(None)` and `Err(LoftyError::UnknownFormat)`?
* Why is the return value `Err(LoftyError::UnknownFormat)` from `FileType::from_buffer_inner` transfomed into `Ok(None)`, but internal errors such as `buf_len < 3` use `Err(LoftyError::UnknownFormat)`?
* Would a return type of `Result<FileType>` make more sense?

Serial-ATA commented:
Thanks for working on this!

The locations of the tests are fine, could you rename the asset to something like `junk.mp3` just to make its purpose a little clearer. Those `a.*` names came from awhile ago, I'll have to update those eventually.

> Why does it return Result<Option<FileType>>?

Since the only error that can occur is `std::io::Error`:

https://github.com/Serial-ATA/lofty-rs/blob/c88cad09b8c0a200eeace10f6e4043fcb2f4c420/src/probe.rs#L138-L142

We just use `Option::or` to set the file type rather than error on no type being found.

> What is the logical difference between the return values Ok(None) and Err(LoftyError::UnknownFormat)?

`Ok(None)` means there were no io errors, but we also couldn't determine the format.

> Why is the return value Err(LoftyError::UnknownFormat) from FileType::from_buffer_inner transfomed into Ok(None), but internal errors such as buf_len < 3 use Err(LoftyError::UnknownFormat)?

That's actually a mistake, the length check shouldn't be there at all. I forgot to remove those when I chose to only return `io::Error`. Good catch. Could you remove the instances of `Err(LoftyError::*)` and just return `Ok(None)` in their place?

Serial-ATA commented on src/mp3/header.rs:
```diff
@@ -9,6 +9,36 @@ pub(crate) fn verify_frame_sync(frame_sync: [u8; 2]) -> bool {
+/// Only the first match is returned and on no match, [`None`] is returned instead.
+///
+/// Note that the search searches in 8 bit steps, i.e. the first 8 bits need to be byte aligned.
+pub(crate) fn search_for_frame_sync(input: &mut dyn Read) -> Result<Option<u64>> {
```
```suggestion
pub(crate) fn search_for_frame_sync<R>(input: &mut R) -> Result<Option<u64>>
where
    R: Read,
{
```

Serial-ATA commented on src/probe.rs:
```diff
@@ -149,41 +149,62 @@ impl<R: Read + Seek> Probe<R> {
+						if let Some(_) = search_for_frame_sync(&mut self.inner)? {
+							Ok(Some(FileType::MP3))
+						} else {
+							Err(LoftyError::UnknownFormat)
```
```suggestion
							Ok(None)
```

Serial-ATA commented on src/probe.rs:
```diff
@@ -247,3 +268,35 @@ where
+		];
+		let data: Vec<u8> = data.into_iter().flatten().cloned().collect();
+		let data = std::io::Cursor::new(&data);
+		let probe = Probe::new(data);
+		let probe = probe.guess_file_type().unwrap();
```
```suggestion
		let probe = Probe::new(data).guess_file_type().unwrap();
```

This can just be one line :)

Serial-ATA commented on src/mp3/read.rs:
```diff
@@ -80,17 +80,30 @@ where
+						start_of_search_area + first_mp3_frame_start_relative;
+
+					// read the first four bytes of the found frame
+					reader.seek(SeekFrom::Start(first_mp3_frame_start_absolute))?;
```
Aren't we already here?

localthomas replied:
No, because the call to `search_for_frame_sync(reader)` changes the reader and the first two bytes of the MP3 frame are consumed. The function needs to read the first 11 bits, i.e. the first two bytes, and those would not be included in the follwing `reader.read_u32`.

It could be replaced by `reader.Seek(SeekFrom::Current(-2))`, but I think this would be less readable and robust.

Serial-ATA replied:
Ah, yeah this is definitely the more readable option. Forgot this will leave the reader past the frame sync.

Serial-ATA commented on src/probe.rs:
```diff
@@ -149,41 +149,62 @@ impl<R: Read + Seek> Probe<R> {
 
 	#[allow(clippy::shadow_unrelated)]
 	fn guess_inner(&mut self) -> Result<Option<FileType>> {
```
```suggestion
	fn guess_inner(&mut self) -> std::io::Result<Option<FileType>> {
```

We can just return an `io::Result` now

Serial-ATA commented on src/probe.rs:
```diff
@@ -149,41 +149,62 @@ impl<R: Read + Seek> Probe<R> {
+						&mut Cursor::new(&mut ident[..]),
+					)?;
+
+					if buf_len < 3 {
+						Err(LoftyError::UnknownFormat)
+					} else if &ident == b"MAC" {
```
```suggestion
					if &ident == b"MAC" {
```

We don't need this check anymore

Serial-ATA reviewed (changes requested):
Looking good! Just some minor things.

localthomas commented:
I have updated the pull request according to the code review and also fixed the clippy errors (very good to hve that in CI).

---

Would it be useful to also change the type definition of `Probe::guess_file_type` to `std::io::Result` to make the comment about "All errors that occur within this function are `std::io::Error`." irrelevant?
It would change the public API, so this might not be desirable to do in this pull request.

Serial-ATA commented:
Yeah, I'll save that change for the next major version. This can be included in the next minor release.

Also, could you revert that clippy lint for the negative multiplication? I think that was a lot more readable. You can just throw an `#[allow(clippy::neg_multiply)]` in there.

Serial-ATA commented:
Thanks!
